core/launcher/src/manager/app_launcher.rs

```diff
@@ -513,8 +515,43 @@ impl AppLauncher {
         }
     }
 
-    fn get_transport(url: String) -> AppRuntimeTransport {
-        if !url.as_str().contains("__firebolt_endpoint") {
+    fn get_modified_url(manifest: AppManifest, sessionid: &str) -> String {
+        let url = manifest.start_page.to_string();
+        if let Ok(mut modified_url) = Url::parse(&url) {
+            let mut query_params: Vec<(String, String)> = modified_url
+                .query_pairs()
+                .map(|(k, v)| (k.into_owned(), v.into_owned()))
+                .collect();
+
+            if !query_params
+                .iter()
+                .any(|(key, _)| key == "__firebolt_endpoint")
+            {
+                let firebolt_endpoint = String::from("127.0.0.1");
+                let appid = manifest.name;
+                let value: String = format!(
+                    "ws://{}:3473?appId={}&session={}",
+                    firebolt_endpoint, appid, sessionid
+                );
+                query_params.push(("__firebolt_endpoint".to_string(), value));
+            }
+            // Modify the URL with the updated query parameters
+            modified_url.set_query(Some(
+                &query_params
+                    .iter()
+                    .map(|(key, value)| format!("{}={}", key, encode(value)))
+                    .collect::<Vec<String>>()
+                    .join("&"),
+            ));
+            modified_url.as_str().to_string()
+        } else {
+            debug!("Invalid URL");
+            "Invalid URL".to_string()
+        }
+    }
+
+    fn get_transport(contract_permited: bool, url: String) -> AppRuntimeTransport {
```

Comment:
url could have been taken as a reference rather than as an owned one. This avoids cloning in the calling place.

Reply:
done